### PR TITLE
Sqlx estimated cost calculation

### PR DESCRIPTION
**Pull request by siddjoshi-ms**
[Preview this branch](https://cosmos-explorer-preview.azurewebsites.net/pull/EDIT_THIS_NUMBER_IN_THE_PR_DESCRIPTION?feature.someFeatureFlagYouMightNeed=true)

Adding cost estimation for Dedicated Gateway Offering (SQLx)

### CODE REVIEW

**Comment by siddjoshi-ms** [comment]
Display cost
image

Update cost on instance count change
image

No screenshot for update cost on sku change as only 1 sku supported right now, but this cost will update whenever sku is changed with the implementation in this PR.

Any error (request error, missing data from fetchPrices API for given regions etc.) leads to default string being shown (default string is what is currently shown)
image

Only number is set when cost is determined to avoid any need for localization in the future.

File: src/SelfServe/SelfServeTypes.ts [hunk 1]
```diff
@@ -236,6 +236,7 @@ export enum DescriptionType {
 export interface Description {
   /**
    * Key used to pickup the string corresponding to the text to be shown as part of the description, from the strings JSON file.
+   * In case key is not present in strings JSON file, displays key as is.
```
Comment (srinathnarayanan) [inline src/SelfServe/SelfServeTypes.ts:239]: Can you remove this. This is not specific to the Description, it is for any translation key. I will add it elsewhere in the documentation

File: src/SelfServe/SqlX/SqlX.rp.ts [hunk 2]
```diff
@@ -128,3 +128,105 @@ export const refreshDedicatedGatewayProvisioning = async (): Promise<RefreshResu
     return { isUpdateInProgress: false, updateInProgressMessageTKey: undefined };
   }
 };
+
+// Data model for Read Regions Query
```
Comment (srinathnarayanan) [inline src/SelfServe/SqlX/SqlX.rp.ts:132]: Remove comments here and everywhere else. I d say add comments only when doing something outside the norm, or if it a todo

File: src/SelfServe/SqlX/SqlX.rp.ts [hunk 3]
```diff
@@ -128,3 +128,105 @@ export const refreshDedicatedGatewayProvisioning = async (): Promise<RefreshResu
...
+    if (response.result.location !== undefined) {
+      readRegions.push(response.result.location.replace(" ", "").toLowerCase());
+    } else {
+      for (let i = 0; i < response.result.locations.length; i++) {
+        readRegions.push(response.result.locations[i].locationName.replace(" ", "").toLowerCase());
+      }
+    }
+    return readRegions;
+  } catch (err) {
+    alert(err);
```
Comment (srinathnarayanan) [inline src/SelfServe/SqlX/SqlX.rp.ts:170]: remove alerting. just throw here, the self serve framework catches and displays error as a error message bar on top of the page, below the command bar

File: src/SelfServe/SqlX/SqlX.rp.ts [hunk 4]
```diff
@@ -128,3 +128,105 @@ export const refreshDedicatedGatewayProvisioning = async (): Promise<RefreshResu
     return { isUpdateInProgress: false, updateInProgressMessageTKey: undefined };
   }
 };
+
+// Data model for Read Regions Query
+interface ReadRegionsResponse {
```
Comment (srinathnarayanan) [inline src/SelfServe/SqlX/SqlX.rp.ts:133]: move all data models to SqlxTypes file

File: src/SelfServe/SqlX/SqlX.rp.ts [hunk 5]
```diff
@@ -128,3 +128,105 @@ export const refreshDedicatedGatewayProvisioning = async (): Promise<RefreshResu
...
+      path: getGeneralPath(userContext.subscriptionId, userContext.resourceGroup, userContext.databaseAccount.name),
+      method: "GET",
+      apiVersion: "2021-04-01-preview",
+    });
+
+    // Parse ARM Response to determine read regions
+    if (response.result.location !== undefined) {
+      readRegions.push(response.result.location.replace(" ", "").toLowerCase());
+    } else {
+      for (let i = 0; i < response.result.locations.length; i++) {
```
Comment (srinathnarayanan) [inline src/SelfServe/SqlX/SqlX.rp.ts:164]: prefer the "for (var location in response.result.locations)" syntax

File: src/SelfServe/SqlX/SqlX.rp.ts [hunk 6]
```diff
@@ -128,3 +128,105 @@ export const refreshDedicatedGatewayProvisioning = async (): Promise<RefreshResu
...
+}
+
+// Query ARM Fetch Prices to get pricing
+export const getPriceMap = async (readRegions: Array<string>): Promise<Map<string, Map<string, number>>> => {
+  try {
+    // Initialize empty PriceMap
+    const priceMap = new Map<string, Map<string, number>>();
+
+    // Query ARM
+    for (let i = 0; i < readRegions.length; i++) {
```
Comment (srinathnarayanan) [inline src/SelfServe/SqlX/SqlX.rp.ts:200]: see above comment on syntax of For Loop

File: src/SelfServe/SqlX/SqlX.rp.ts [hunk 7]
```diff
@@ -128,3 +128,105 @@ export const refreshDedicatedGatewayProvisioning = async (): Promise<RefreshResu
...
+        queryParams: {
+          filter:
+            "armRegionName eq " +
+            readRegions[i] +
+            " and serviceFamily eq 'Databases' and productName eq 'Azure Cosmos DB Dedicated Gateway - General Purpose'",
+        },
+      });
+
+      // Parse response to create price mapping for current region
+      for (let j = 0; j < response.result.Items.length; j++) {
```
Comment (srinathnarayanan) [inline src/SelfServe/SqlX/SqlX.rp.ts:218]: see above comment on syntax of For Loop

File: src/SelfServe/SqlX/SqlX.rp.ts [hunk 8]
```diff
@@ -128,3 +128,105 @@ export const refreshDedicatedGatewayProvisioning = async (): Promise<RefreshResu
...
+        regionPriceMap.set(response.result.Items[j].skuName, response.result.Items[j].retailPrice);
+      }
+
+      // Insert region's PriceMap to overall PriceMap
+      priceMap.set(readRegions[i], regionPriceMap);
+    }
+
+    return priceMap;
+  } catch (err) {
+    alert(err);
```
Comment (srinathnarayanan) [inline src/SelfServe/SqlX/SqlX.rp.ts:228]: remove alert. refer getDedicatedGatewayResource for throwing on error after logging telemetry.

File: src/SelfServe/SqlX/SqlX.tsx [hunk 9]
```diff
@@ -79,6 +86,12 @@ const onNumberOfInstancesChange = (
   } else {
     currentValues.set("warningBanner", undefined);
   }
+
+  // Update cost per hour based on new instance count
```
Comment (srinathnarayanan) [inline src/SelfServe/SqlX/SqlX.tsx:90]: Remove comments here and everywhere else

File: src/SelfServe/SqlX/SqlX.tsx [hunk 10]
```diff
@@ -177,6 +190,27 @@ const NumberOfInstancesDropdownInfo: Info = {
   },
 };
 
+// Block-scoped Variables to store information for cost calculation
+let priceMap: Map<string, Map<string, number>>;
```
Comment (srinathnarayanan) [inline src/SelfServe/SqlX/SqlX.tsx:194]: why declare it here? Declare and use where it is initialized

File: src/SelfServe/SqlX/SqlX.tsx [hunk 11]
```diff
@@ -339,7 +377,7 @@ export default class SqlX extends SelfServeBaseClass {
   instances: number;
 
   @Values({
-    labelTKey: "Cost",
+    labelTKey: "Approximate Cost Per Hour",
```
Comment (srinathnarayanan) [inline src/SelfServe/SqlX/SqlX.tsx:380]: Add entry for "Approximate Cost Per Hour" in SqlX.json file 

example 
{
...
...
"ApproximateCost": "Approximate Cost Per Hour",
...
...
}

File: src/SelfServe/SqlX/SqlX.tsx [hunk 12]
```diff
@@ -177,6 +190,27 @@ const NumberOfInstancesDropdownInfo: Info = {
   },
 };
 
+// Block-scoped Variables to store information for cost calculation
+let priceMap: Map<string, Map<string, number>>;
+let readRegions: Array<string>;
+
+const calculateCost = (skuName: string, instanceCount: number): string | Description => {
```
Comment (srinathnarayanan) [inline src/SelfServe/SqlX/SqlX.tsx:197]: this only returns a description right?

File: src/SelfServe/SqlX/SqlX.tsx [hunk 13]
```diff
@@ -53,7 +55,12 @@ const CosmosD16s = "Cosmos.D16s";
 
 const onSKUChange = (newValue: InputType, currentValues: Map<string, SmartUiInput>): Map<string, SmartUiInput> => {
   currentValues.set("sku", { value: newValue });
-  currentValues.set("costPerHour", { value: costPerHourValue });
+
+  // Update cost per hour based on new sku
+  currentValues.set("costPerHour", {
+    value: calculateCost(newValue as string, currentValues.get("NumberOfInstances").value as number),
```
Comment (srinathnarayanan) [inline src/SelfServe/SqlX/SqlX.tsx:61]: this should be "instances" right?
This scenario is broken :
Account does not have sqlx provisioned, go to Dedicated gateway blade, change the Sku type.

File: src/SelfServe/SqlX/SqlX.tsx (hunk 10 above)
Comment (siddjoshi-ms) [reply src/SelfServe/SqlX/SqlX.tsx:194]: Need this to be available outside the function that initializes it, that's why have declared it outside the scope of the initalize function.

**Comment by siddjoshi-ms** [comment]
Info bubble added

image

File: src/Localization/en/SqlX.json [hunk 14]
```diff
@@ -40,7 +40,7 @@
     "CosmosD4Details": "General Purpose Cosmos Compute with 4 vCPUs, 16 GB Memory",
     "CosmosD8Details": "General Purpose Cosmos Compute with 8 vCPUs, 32 GB Memory",
     "CosmosD16Details": "General Purpose Cosmos Compute with 16 vCPUs, 64 GB Memory",
-    "Cost": "Cost",
+    "ApproximateCost": "Approximate Cost Per Hour",
```
Comment (timsander1) [inline src/Localization/en/SqlX.json:43]: Could we mention somewhere that the cost is in USD? Right now, we just have the $ in the cost estimate which could be ambiguous.

Comment (siddjoshi-ms) [reply src/Localization/en/SqlX.json:43]: Made it US $, does that work?

Comment (timsander1) [reply src/Localization/en/SqlX.json:43]: yeah that works!

File: src/SelfServe/SqlX/SqlX.tsx [hunk 15]
```diff
@@ -79,6 +84,11 @@ const onNumberOfInstancesChange = (
   } else {
     currentValues.set("warningBanner", undefined);
   }
+
+  currentValues.set("costPerHour", {
```
Comment (srinathnarayanan) [inline src/SelfServe/SqlX/SqlX.tsx:88]: set costPerHour in the onEnableDedicatedGatewayChange function as well, based on whether 'newValue' is true or false. calculate in the true case, hide in false case.

File: src/SelfServe/SqlX/SqlX.tsx [hunk 16]
```diff
@@ -79,6 +84,11 @@ const onNumberOfInstancesChange = (
...
+    value: calculateCost(currentValues.get("sku").value as string, newValue as number),
```
Comment (srinathnarayanan) [inline src/SelfServe/SqlX/SqlX.tsx:89]: limit to 2 decimals

File: src/SelfServe/SqlX/SqlX.tsx [hunk 17]
```diff
@@ -92,7 +102,19 @@ const onEnableDedicatedGatewayChange = (
   if (dedicatedGatewayOriginallyEnabled === newValue) {
     currentValues.set("sku", baselineValues.get("sku"));
     currentValues.set("instances", baselineValues.get("instances"));
-    currentValues.set("costPerHour", baselineValues.get("costPerHour"));
+
+    if (newValue === true) {
```
Comment (srinathnarayanan) [inline src/SelfServe/SqlX/SqlX.tsx:106]: there is already a 'if newValue === true' block. Move this there

File: src/SelfServe/SqlX/SqlX.tsx [hunk 18]
```diff
@@ -274,12 +336,16 @@ export default class SqlX extends SelfServeBaseClass {
       hidden: true,
     });
 
+    // Get Read Regions and PriceMap
```
Comment (srinathnarayanan) [inline src/SelfServe/SqlX/SqlX.tsx:339]: remove comment

File: src/SelfServe/SqlX/SqlX.tsx (hunk 17 above)
Comment (siddjoshi-ms) [reply src/SelfServe/SqlX/SqlX.tsx:106]: the path inside this if statement will not hit the other if newValue===true block, it returns at the end of the if statement.
I can reorganize the code to only check if newValue===true once, but I wasn't sure if that was acceptable, so just made minimal change to existing code.

Comment (srinathnarayanan) [reply src/SelfServe/SqlX/SqlX.tsx:106]: dedicatedGatewayOriginallyEnabled === newValue ====> set everything to the baseline. it is already computed on init. Set costPerHour to baseline as well. 

This newValue ===true logic is irrespective of whether dedicatedGatewayOriginallyEnabled === newValue or not. So move it to that section

**Review by srinathnarayanan — APPROVED** [review]

**Review by southpolesteve — APPROVED** [review]

**Pull request merged**